weblate/api/views.py

```diff
@@ -752,6 +755,141 @@ def destroy(self, request, *args, **kwargs):
+
+            project_slug = request.data["project_slug"]
+
+            try:
+                project = Project.objects.get(slug=project_slug)
+            except Project.DoesNotExist:
+                raise Http404("No project slug '%s' found!" % project_slug)
+
+            obj.links.add(project)
+            serializer = self.serializer_class(obj, context={"request": request})
+
+            return Response(data={"data": serializer.data}, status=HTTP_201_CREATED)
+
+        queryset = obj.links.order_by("id")
+        page = self.paginate_queryset(queryset)
+
+        serializer = ProjectSerializer(page, many=True, context={"request": request})
+
+        return self.get_paginated_response(serializer.data)
+
+    @action(
+        detail=True, methods=["delete"], url_path="projects/(?P<project_slug>[^/.]+)"
+    )
+    def delete_projects(self, request, id, project_slug):
+        obj = self.get_object()
+        self.perm_check(request)
+
+        try:
+            project = Project.objects.get(slug=project_slug)
+        except (Project.DoesNotExist, ValueError) as error:
+            return Response(
+                data={"result": "Unsuccessful", "detail": force_str(error)},
+                status=HTTP_400_BAD_REQUEST,
+            )
+        obj.links.remove(project)
+        return Response(status=HTTP_204_NO_CONTENT)
+
+    @action(detail=True, methods=["get", "post"], serializer_class=TermSerializer)
+    def terms(self, request, **kwargs):
+        obj = self.get_object()
+        if request.method == "POST":
+            if not request.user.has_perm("glossary.add"):
+                self.permission_denied(request, message="Can not create terms")
+            with transaction.atomic():
+                serializer = TermSerializer(
+                    data=request.data, context={"request": request}
+                )
+                serializer.is_valid(raise_exception=True)
+                serializer.save(glossary=obj, user=request.user)
+                return Response(serializer.data, status=HTTP_201_CREATED)
+
+        queryset = obj.term_set.order_by("id")
+        page = self.paginate_queryset(queryset)
+
+        serializer = TermSerializer(page, many=True, context={"request": request})
+
+        return self.get_paginated_response(serializer.data)
+
+    @action(
+        detail=True,
+        methods=["get", "put", "patch", "delete"],
+        url_path="terms/(?P<term_id>[^/.]+)",
+        serializer_class=TermSerializer,
+    )
+    def terms_details(self, request, id, term_id):
+        obj = self.get_object()
+
+        try:
+            term = obj.term_set.get(id=term_id)
+        except (Term.DoesNotExist, ValueError) as error:
+            return Response(
+                data={"result": "Unsuccessful", "detail": force_str(error)},
+                status=HTTP_400_BAD_REQUEST,
+            )
+
+        if request.method == "DELETE":
+            if not request.user.has_perm("glossary.delete"):
```

Comment:
You need to pass matching project object here as well. In the views this is already scoped to a project, so we use that one, but here we should probably check all glossary projects. Something like (not tested):

```suggestion
            if not any(request.user.has_perm("glossary.delete", project) for project in obj.projects.all()):
```

Same applies to other `glossary.*` permission checks.